TASK-12: discovery page tabs

r-saba opened the pull request:
This PR implements the discovery page tabs #54.

To test the component visit the followin URL:
`http://localhost:4200/home`

Swiping between routes is possible when in mobile view

r-saba committed:
Add subnavigation menu without swipe functionality #54

r-saba committed:
Add swipeable routes, failing tests #54

r-saba committed:
Remove snapshot tests for app.tsx and change routing logic #54

r-saba committed:
Add tests for subnavigation active buttons #54

r-saba committed:
Add specific routes for the subnavigation, make /discover the initial page shown #54

r-saba committed:
refactor: Remove unecessary variables from existing tests #54

cristian-aldea commented on package.json:
```diff
@@ -69,6 +69,7 @@
     "react-dom": "16.13.1",
     "react-nice-dates": "^3.0.7",
     "react-router-dom": "5.2.0",
+    "react-swipeable-routes": "^0.8.0",
```
I'm getting the following warning in the console, which seems to be coming from the ReactSwipableView:
```
Warning: Legacy context API has been detected within a strict-mode tree.

The old API will be supported in all 16.x releases, but applications using it should migrate to the new version.

Please update the following components: ReactSwipableView

Learn more about this warning here: https://fb.me/react-legacy-context
    in ReactSwipableView (created by SwipeableRoutes)
    in SwipeableRoutes (created by Context.Consumer)
    in withRouter(SwipeableRoutes) (created by App)
    in Route (created by App)
    in Switch (created by App)
    in div (created by App)
    in App
    in Router (created by BrowserRouter)
    in BrowserRouter
    in ApolloProvider
    in StrictMode
printWarning @ vendor.js:127606
vendor.js:127606 

Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-unsafe-component-lifecycles for details.

* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
* Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.

Please update the following components: ReactSwipableView
```

Would there be a way for us to implement this sliding ourselves? Or update the package to not have this warning?

r-saba replied:
This issue is currently present in almost all UI Framworks ([materia-ui](https://github.com/mui-org/material-ui/issues/13394), [react-strap](https://github.com/reactstrap/reactstrap/issues/1833)). Users have not reported any issues as a result of this warning.

Also [react-swipeable-views](https://github.com/oliviertassinari/react-swipeable-views) is widely used so I wouldn't worry about support for this package. It's used by Uber, Tinder and Material-UI. The warning in [known](https://github.com/oliviertassinari/react-swipeable-views/issues/596).

Considering all of this I don't see the benefit of making our own implementation.

cristian-aldea replied:
Alright makes sense, just wanted a bit more insight into this.

cristian-aldea commented on apps/mull-ui/src/app/app.scss:
```diff
@@ -22,3 +22,9 @@
     padding: $nav-bar-height-desktop 0 0;
   }
 }
+
+// Global Styles
+
+a {
+  text-decoration: none;
+}
```
This could be moved to `libs/ui-lib/src/lib/styles/styles.scss`

r-saba replied:
I put it here because we're getting rid of ui-lib but yeah I can put it there for now and we can move it all at once

cristian-aldea replied:
yeah for now keep it in there, we'll move it later

cristian-aldea commented on apps/mull-ui/src/app/app.tsx:
```diff
@@ -41,6 +44,14 @@ export const App = () => {
             <EventPage event={dummyEvent} prevPage={'Review'} />
           </div>
         </Route>
+        {/* Without this outter Route component, the inner components would show on all routes that are not matched */}
```
I think it would be better if we kept the home page and made the three pages you added be subpages, so something like:
- `/home/discover`
- `/home/upcoming`
- `/home/myevents`

What do you think @r-saba?

r-saba replied:
That's how I had it, initially I preferred it without the `/home`  since it with concise

cristian-aldea replied:
considering we have a button for the home page id say the way you had it initially was good. Having them under a common base route of `/home` makes it easier to understand that its still the home page, but just different tabs.

cristian-aldea replied:
I don't mind either way, so I won't block over it.

cristian-aldea commented on apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx:
```diff
@@ -0,0 +1,39 @@
+import { ROUTES } from 'apps/mull-ui/src/constants';
+import React from 'react';
+import { NavLink, useRouteMatch } from 'react-router-dom';
```
```suggestion
import { NavLink } from 'react-router-dom';
```
Make sure to lint your code before making a pr. There are other linting warnings/errors I found from your changes, so please fix them.

cristian-aldea commented on apps/mull-ui/src/app/app.tsx:
```diff
@@ -41,6 +44,14 @@ export const App = () => {
           </div>
         </Route>
+        {/* Without this outter Route component, the inner components would show on all routes that are not matched */}
+        <Route exact path={[`${ROUTES.DISCOVER}`, `${ROUTES.UPCOMING}`, `${ROUTES.MYEVENTS}`]}>
+          <SwipeableRoutes>
+            <Route path={ROUTES.DISCOVER} component={SubNavigationBar} />
+            <Route path={ROUTES.UPCOMING} component={SubNavigationBar} />
+            <Route path={ROUTES.MYEVENTS} component={SubNavigationBar} />
+          </SwipeableRoutes>
+        </Route>
```
I don't know about you @r-saba but I really don't like how the sub-navbar slides along with the page. It makes it really confusing. I experimented a bit and found a solution. Let me know what you think.
```suggestion
        <Route exact path={[`${ROUTES.DISCOVER}`, `${ROUTES.UPCOMING}`, `${ROUTES.MYEVENTS}`]}>
          <SubNavigationBar />
          <SwipeableRoutes>
            <Route path={ROUTES.DISCOVER} component={() => <div>DISCOVER!</div>} />
            <Route path={ROUTES.UPCOMING} component={() => <div>UPCOMING!</div>} />
            <Route path={ROUTES.MYEVENTS} component={() => <div>MYEVENTS!</div>} />
          </SwipeableRoutes>
        </Route>
```

r-saba replied:
Oh good point, didn't consider that we probably don't want the subnavigation to slide as well

cristian-aldea reviewed (changes requested):
Nice implementation @r-saba! I had a few concerns and comments I'd like some clarifications on, but otherwise nice job.

r-saba committed:
Address pr comments #54

cristian-aldea reviewed (commented):
Thanks for addressing my comments, I don't have time to re-review, and I don't want to block this pr, so I'll approve it immediately if it gets another approval. I'll be re-reviewing in more detail tomorrow.

bit172 commented on apps/mull-ui/src/constants.ts:
```diff
@@ -4,14 +4,16 @@ import { EventRestriction, IEvent } from '@mull/types';
   REGISTER: '/register',
+  DISCOVER: '/discover',
+  UPCOMING: '/upcoming',
+  MYEVENTS: '/myevents',
```
This is really minor but I think it should be consistent
```suggestion
  MY_EVENTS: '/my-events',
```

bit172 replied:
This is really minor but I think it should be consistent
```suggestion
  MY_EVENTS: '/my-events',
```

bit172 reviewed (changes requested):
Looks very good! I just have a really minor change. Good job!

JimLau49 reviewed (dismissed):
Looks good to me 🚀  . I think it should have been mentioned that to test scrollability that you need to click along the length of container that has "DISCOVER!" "UPCOMING!" or "MYEVENTS!"
**Acceptance Criteria**

- [x] The tabs are implemented for the discovery page and the user can swipe/tap to switch between tabs.
- [x] Each tab is routed, but can lead to dead links or pages
- [x] Documentation present for every new component and function
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=533%3A6528)

cristian-aldea commented on apps/mull-ui/src/app/app.tsx:
```diff
@@ -41,6 +44,15 @@ export const App = () => {
           </div>
         </Route>
+        {/* Without this outter Route component, the inner components would show on all routes that are not matched */}
+        <Route exact path={[`${ROUTES.DISCOVER}`, `${ROUTES.UPCOMING}`, `${ROUTES.MYEVENTS}`]}>
```
Is there a reason why we can't have:
```suggestion
        <Route exact path={[ROUTES.DISCOVER, ROUTES.UPCOMING, ROUTES.MYEVENTS]}>
```
I tried it and it seems to work

cristian-aldea commented on apps/mull-ui/src/app/app.tsx:
```diff
@@ -41,6 +44,15 @@ export const App = () => {
           </div>
         </Route>
+        {/* Without this outter Route component, the inner components would show on all routes that are not matched */}
+        <Route exact path={[`${ROUTES.DISCOVER}`, `${ROUTES.UPCOMING}`, `${ROUTES.MYEVENTS}`]}>
+          <SubNavigationBar />
+          <SwipeableRoutes>
+            <Route path={ROUTES.DISCOVER} component={() => <div>DISCOVER!</div>} />
+            <Route path={ROUTES.UPCOMING} component={() => <div>UPCOMING!</div>} />
+            <Route path={ROUTES.MYEVENTS} component={() => <div>MYEVENTS!</div>} />
+          </SwipeableRoutes>
+        </Route>
```
One issue im seeing right now with styling is that the Subnavbar is set to be in the page-container, which is then pushing the actual page content further down than it should.

![image](https://user-images.githubusercontent.com/23544999/100519347-5015dc80-3165-11eb-8845-e3c93d157709.png)

I propose this solution: Remove the page-container class from SubNavigationBar, and put page-container on a container div. So something like this:
```suggestion
        <Route exact path={[ROUTES.DISCOVER, ROUTES.UPCOMING, ROUTES.MYEVENTS]}>
          <div className="page-container">
            <SubNavigationBar />
            <SwipeableRoutes>
              <Route path={ROUTES.DISCOVER} component={() => <div>DISCOVER!</div>} />
              <Route path={ROUTES.UPCOMING} component={() => <div>UPCOMING!</div>} />
              <Route path={ROUTES.MYEVENTS} component={() => <div>MYEVENTS!</div>} />
            </SwipeableRoutes>
          </div>
        </Route>
```

cristian-aldea commented on apps/mull-ui/src/app/app.tsx:
```diff
@@ -41,6 +44,15 @@ export const App = () => {
           </div>
         </Route>
+        {/* Without this outter Route component, the inner components would show on all routes that are not matched */}
+        <Route exact path={[`${ROUTES.DISCOVER}`, `${ROUTES.UPCOMING}`, `${ROUTES.MYEVENTS}`]}>
```
The more I think about it the more I think these routes should be under `/home`, since each tab is a subpage for the home page. I'm not gonna block the PR over this, but I really think what you did initially was good.

r-saba replied:
Sure, I will change it

cristian-aldea commented on apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.tsx:
```diff
@@ -0,0 +1,39 @@
+
+import './subnavigation-bar.scss';
+
+export const SubNavigationBar = () => {
```
Add this to the `apps/mull-ui/src/app/components/index.ts` 👀

r-saba replied:
you do a good job of catching these minor details!

cristian-aldea replied:
More 👀 more 💵

cristian-aldea commented on apps/mull-ui/src/app/components/subnavigation-bar/subnavigation-bar.spec.tsx:
```diff
@@ -0,0 +1,55 @@
+  it('should have the correct button active based on the url', () => {
+    const testIds = {
+      [ROUTES.DISCOVER]: 'subnavigation-discover-button',
+      [`${ROUTES.UPCOMING}`]: 'subnavigation-upcoming-button',
+      [`${ROUTES.MYEVENTS}`]: 'subnavigation-myEvents-button',
```
Same comment as in app.tsx, is there a reason we can't put:
```suggestion
      [ROUTES.UPCOMING]: 'subnavigation-upcoming-button',
      [ROUTES.MYEVENTS]: 'subnavigation-myEvents-button',
```
I tried it and the tests still seem to work.

r-saba replied:
It's a remnant from my first iteration, I did a simple find and replace, should be what you proposed now.

cristian-aldea reviewed (changes requested):
I had a chance to look at your pr in more detail and I found a few more issues I'd like addressed or replied to.

r-saba committed:
Address more pr comments #54

r-saba commented:
I addressed your pr comments @bit172 @TheGreatMarkus @JimLau49 . Let me know how it's looking.
Major difference is now each tab points to a nested route `/home/:id`

cristian-aldea commented on apps/mull-ui/src/app/app.tsx:
```diff
@@ -41,6 +43,16 @@ export const App = () => {
+          <div className="page-container">
+            <SubNavigationBar />
+            <SwipeableRoutes>
+              <Route path={ROUTES.DISCOVER} component={() => <div>DISCOVER!</div>} />
+              <Route path={ROUTES.UPCOMING} component={() => <div>UPCOMING!</div>} />
+              <Route path={ROUTES.MY_EVENTS} component={() => <div>MY_EVENTS!</div>} />
```
I got one last suggestion here. This is to avoid people routing to /home/somethingelse
```suggestion
              <Route path={ROUTES.DISCOVER} component={() => <div>DISCOVER!</div>} />
              <Route path={ROUTES.UPCOMING} component={() => <div>UPCOMING!</div>} />
              <Route path={ROUTES.MY_EVENTS} component={() => <div>MY_EVENTS!</div>} />
              <Route path={ROUTES.HOME} render={() => <Redirect to={ROUTES.DISCOVER} />} />
```

cristian-aldea reviewed (dismissed):
Great Work Ragith! I just had one last issue I saw. But otherwise this is as good as merged.

- [x] The tabs are implemented for the discovery page and the user can swipe/tap to switch between tabs.
- [x] Each tab is routed but can lead to dead links or pages
- [x] Documentation present where needed
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=533%3A6528)

bit172 reviewed (dismissed):
LGTM! 🚀

- [x] The tabs are implemented for the discovery page and the user can swipe/tap to switch between tabs.
- [x] Each tab is routed but can lead to dead links or pages
- [x] Documentation present where needed
- [x] Unit and UI tests are implemented for every component and function
- [x] Must match [Mockup](https://www.figma.com/file/FJPXMYI4ULCnuCEZvPs67k/Mull?node-id=533%3A6528)

r-saba committed:
Add redirect for all /home/* routes #54

r-saba committed:
Fix prettier issue with pr #54
